File: src/1203_sort_items_by_groups_respecting_dependencies.js
class ListNode {
  constructor(val = 0, next = null) { this.val = val; this.next = next; }
}

class TreeNode {
  constructor(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
}

class Node {
  constructor(val = 0, next = null, random = null, left = null, right = null, neighbors = null, key = -1) {
    this.val = val;
    this.next = next;
    this.random = random;
    this.left = left;
    this.right = right;
    this.neighbors = neighbors === null ? [] : neighbors;
    this.key = key;
  }
}

function listNodeToString(node) {
  const parts = [];
  for (let cur = node; cur !== null; cur = cur.next) parts.push(String(cur.val));
  return parts.length ? parts.join('->') : 'None';
}

function treeToString(node) {
  if (node === null || node === undefined) return 'None';
  return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
}

function buildListNode(values) {
  const dummy = new ListNode();
  let tail = dummy;
  for (const value of values) {
    tail.next = new ListNode(value);
    tail = tail.next;
  }
  return dummy.next;
}

class MinHeap {
  constructor(compare = (a, b) => a < b) { this.data = []; this.compare = compare; }
  get length() { return this.data.length; }
  peek() { return this.data[0]; }
  push(value) {
    this.data.push(value);
    this._up(this.data.length - 1);
  }
  pop() {
    if (this.data.length === 0) return undefined;
    const root = this.data[0];
    const last = this.data.pop();
    if (this.data.length > 0) {
      this.data[0] = last;
      this._down(0);
    }
    return root;
  }
  _up(i) {
    while (i > 0) {
      const p = Math.floor((i - 1) / 2);
      if (!this.compare(this.data[i], this.data[p])) break;
      [this.data[i], this.data[p]] = [this.data[p], this.data[i]];
      i = p;
    }
  }
  _down(i) {
    while (true) {
      const l = i * 2 + 1;
      const r = l + 1;
      let best = i;
      if (l < this.data.length && this.compare(this.data[l], this.data[best])) best = l;
      if (r < this.data.length && this.compare(this.data[r], this.data[best])) best = r;
      if (best === i) break;
      [this.data[i], this.data[best]] = [this.data[best], this.data[i]];
      i = best;
    }
  }
}

class UnionFind {
  constructor(n = null) {
    this.f = new Map();
    this.par = n === null ? [] : Array.from({ length: n }, (_, i) => i);
    this.rank = n === null ? [] : Array(n).fill(1);
    this.size = n === null ? [] : Array(n).fill(1);
    this.count = n === null ? 0 : n;
  }
  findParent(x) {
    const y = this.f.has(x) ? this.f.get(x) : x;
    if (x !== y) {
      const root = this.findParent(y);
      this.f.set(x, root);
      return root;
    }
    return y;
  }
  find(x) {
    if (this.par.length === 0) return this.findParent(x);
    while (x !== this.par[x]) {
      this.par[x] = this.par[this.par[x]];
      x = this.par[x];
    }
    return x;
  }
  union(x, y) {
    if (this.par.length === 0) {
      const px = this.findParent(x), py = this.findParent(y);
      if (px === py) return false;
      this.f.set(px, py);
      return true;
    }
    const px = this.find(x), py = this.find(y);
    if (px === py) return false;
    if (this.rank[px] > this.rank[py]) {
      this.par[py] = px;
      this.rank[px] += this.rank[py];
      this.size[px] += this.size[py];
    } else {
      this.par[px] = py;
      this.rank[py] += this.rank[px];
      this.size[py] += this.size[px];
    }
    this.count -= 1;
    return true;
  }
}

function topologicalSort(successors, predecessorsCount, numNodes) {
  const order = [];
  const nodesWithNoPredecessors = [];
  for (let node = 0; node < numNodes; node++) if (!predecessorsCount[node]) nodesWithNoPredecessors.push(node);
  while (nodesWithNoPredecessors.length) {
    const node = nodesWithNoPredecessors.shift();
    order.push(node);
    for (const successor of successors[node]) {
      predecessorsCount[successor] -= 1;
      if (!predecessorsCount[successor]) nodesWithNoPredecessors.push(successor);
    }
  }
  return order.length === numNodes ? order : [];
}
function sortItems(n, m, group, beforeItems) {
  for (let item = 0; item < n; item++) if (group[item] === -1) group[item] = m++;
  const successorsGroup = Array.from({ length: m }, () => []), successorsItem = Array.from({ length: n }, () => []);
  const predGroup = Array(m).fill(0), predItem = Array(n).fill(0);
  for (let item = 0; item < n; item++) {
    const currentGroup = group[item];
    for (const before of beforeItems[item]) {
      const beforeGroup = group[before];
      if (currentGroup === beforeGroup) { successorsItem[before].push(item); predItem[item] += 1; }
      else { successorsGroup[beforeGroup].push(currentGroup); predGroup[currentGroup] += 1; }
    }
  }
  const groupsOrder = topologicalSort(successorsGroup, predGroup, m);
  const itemsOrder = topologicalSort(successorsItem, predItem, n);
  if (!groupsOrder.length || !itemsOrder.length) return [];
  const itemsGrouped = Array.from({ length: m }, () => []);
  for (const item of itemsOrder) itemsGrouped[group[item]].push(item);
  const result = [];
  for (const grp of groupsOrder) result.push(...itemsGrouped[grp]);
  return result;
}

module.exports = { topologicalSort, sortItems };

if (require.main === module) {
  const assert = require('assert');
  assert.deepStrictEqual(topologicalSort([[0]], [0], 0), []);
}

File: src/1203_sort_items_by_groups_respecting_dependencies.rs
use std::collections::VecDeque;

struct Solution;

impl Solution {
    pub fn topological_sort(
        successors: Vec<Vec<i32>>,
        mut predecessors_count: Vec<i32>,
        num_nodes: i32,
    ) -> Vec<i32> {
        let mut order = Vec::new();
        let mut nodes_with_no_predecessors = VecDeque::new();
        for node in 0..num_nodes as usize {
            if predecessors_count[node] == 0 {
                nodes_with_no_predecessors.push_back(node);
            }
        }

        while let Some(node) = nodes_with_no_predecessors.pop_front() {
            order.push(node as i32);
            for &successor in &successors[node] {
                let successor = successor as usize;
                predecessors_count[successor] -= 1;
                if predecessors_count[successor] == 0 {
                    nodes_with_no_predecessors.push_back(successor);
                }
            }
        }

        if order.len() == num_nodes as usize {
            order
        } else {
            Vec::new()
        }
    }

    #[allow(dead_code)]
    pub fn sort_items(
        n: i32,
        mut m: i32,
        mut group: Vec<i32>,
        before_items: Vec<Vec<i32>>,
    ) -> Vec<i32> {
        for item in 0..n as usize {
            if group[item] == -1 {
                group[item] = m;
                m += 1;
            }
        }

        let mut successors_group = vec![Vec::new(); m as usize];
        let mut successors_item = vec![Vec::new(); n as usize];
        let mut predecessors_count_group = vec![0; m as usize];
        let mut predecessors_count_item = vec![0; n as usize];

        for item in 0..n as usize {
            let current_group = group[item];
            for &before in &before_items[item] {
                let before = before as usize;
                let before_group = group[before];
                if current_group == before_group {
                    successors_item[before].push(item as i32);
                    predecessors_count_item[item] += 1;
                } else {
                    successors_group[before_group as usize].push(current_group);
                    predecessors_count_group[current_group as usize] += 1;
                }
            }
        }

        let groups_order = Self::topological_sort(successors_group, predecessors_count_group, m);
        let items_order = Self::topological_sort(successors_item, predecessors_count_item, n);

        if groups_order.is_empty() || items_order.is_empty() {
            return Vec::new();
        }

        let mut items_grouped = vec![Vec::new(); m as usize];
        for item in items_order {
            items_grouped[group[item as usize] as usize].push(item);
        }

        let mut result = Vec::new();
        for grp in groups_order {
            result.extend(items_grouped[grp as usize].iter().copied());
        }
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::topological_sort(vec![vec![0]], vec![0], 0),
            Vec::<i32>::new()
        );
    }
}
